lib/spack/spack/binary_distribution.py

```diff
@@ -1407,19 +1407,29 @@ def needs_rebuild(spec, mirror_url, rebuild_on_errors=False):
     # did, but didn't match the one we computed locally, then we should
     # just rebuild.  This can be simplified once the dag_hash and the
     # full_hash become the same thing.
-    if ('full_hash' not in spec_yaml or
-            spec_yaml['full_hash'] != pkg_full_hash):
-        if 'full_hash' in spec_yaml:
+    yaml_spec = spec_yaml['spec']
+    name = spec.name
+    cached_pkg_specs = [item[name] for item in yaml_spec if name in item]
+    cached_target = cached_pkg_specs[0] if cached_pkg_specs else None
```

Comment:
This list comprehension is doing a lot of heavy lifting. Can you add a comment to explain why it works for future maintainability. At first glance it's not clear why you're using a list at all, although I think I've convinced myself it's necessary.

Reply:
I don't think I know why the `spec` in the buildcache `spec.yaml` file is a list.  I know I wrote the schema for that thing not too long ago, but I just made it match the pre-existing structure I saw in those files already.  In the database, `spec` is instead an object, with what appears to ever only be a single key, the package name.

@gartung Were you involved in coming up with the buildcache `spec.yaml` file format?  Or are you otherwise aware of any reason the `spec` in there should be a list?  As far as I know, it's supposed to correspond to the binary package for a single concrete spec.

At any rate, because it's a list, I took the approach above to get my hands on the spec so I could get at the `full_hash`.  I've only ever seen that list have a single element, so I guess I could also just assume there will ever be only one element and just grab the first one.

@becker33 Let me know if you want me to rework this a bit, otherwise I'll just add a comment summarizing the above.

Reply:
The comment looks good. I was fine with the code, it just took long enough to parse that it warranted a comment.

Reply:
Sorry I don't recall.

Reply:
As far as I recall, the spec is taken from spec.yaml in .spaclk for the patckage
https://github.com/spack/spack/blob/7cfdf61979bdabb15cfeedb2b505b7ef3725bf28/lib/spack/spack/binary_distribution.py#L702

Reply:
Then the buildcache tarball checksum and whether it uses relative rpaths is added as dictionary entries to the spec
https://github.com/spack/spack/blob/7cfdf61979bdabb15cfeedb2b505b7ef3725bf28/lib/spack/spack/binary_distribution.py#L765
 I am not sure why it is being interpreted as a list. Maybe my understanding of how yaml files are interpreted was incorrect.